### PR TITLE
Adding unsafe methods to the RootNode, InternalNode and LeafNodes

### PR DESCRIPTION
**Pull request by danrbailey**
This adds `isValueOff()` to InternalNode and LeafNode(s) and `probeValue(Index, val)` to LeafNodeBool/LeafNodeMask.

In addition this adds "unsafe" index-based methods to the InternalNode and a few equivalent methods to the LeafNode(s). All these methods have full unit test coverage, below is the complete list of new methods:

**Root Node**

```
RootNode::getValueUnsafe()
RootNode::getChildUnsafe()
RootNode::getConstChildUnsafe()
```

**InternalNode**

```
InternalNode::isValueOff()
InternalNode::getValueUnsafe()
InternalNode::getChildUnsafe()
InternalNode::getConstChildUnsafe()
InternalNode::setActiveStateUnsafe()
InternalNode::setValueOnlyUnsafe()
InternalNode::setValueOnUnsafe()
InternalNode::setValueOffUnsafe()
InternalNode::setChildUnsafe()
InternalNode::resetChildUnsafe()
InternalNode::stealChildUnsafe()
InternalNode::deleteChildUnsafe()
```

**LeafNode (+ LeafNodeMask, LeafNodeBool)**

```
LeafNode::isValueOff()
LeafNode::probeValue(Index,val)
LeafNode::getValueUnsafe()
LeafNode::setActiveStateUnsafe()
LeafNode::setValueOnlyUnsafe()
LeafNode::setValueOnUnsafe()
LeafNode::setValueOffUnsafe()
```

### CODE REVIEW

File: openvdb/openvdb/tree/InternalNode.h [hunk 1]
```diff
@@ -458,6 +462,79 @@ class InternalNode
     void readBuffers(std::istream&, const CoordBBox&, bool fromHalf = false);
 
 
+    //
+    // Unsafe methods
+    //
+    // WARNING: For improved performance, these unsafe methods do not check the value or
```
Comment (kmuseth) [inline openvdb/openvdb/tree/InternalNode.h:468]: I would add this comment to each relevant method with `/// @warning`

Comment (kmuseth) [reply openvdb/openvdb/tree/InternalNode.h:468]: oh I see you already did that so never mind

File: openvdb/openvdb/tree/InternalNode.h [hunk 2]
```diff
@@ -1554,6 +1631,15 @@ InternalNode<ChildT, Log2Dim>::isValueOn(const Coord& xyz) const
     return mNodes[n].getChild()->isValueOn(xyz);
 }
 
+template<typename ChildT, Index Log2Dim>
+inline bool
+InternalNode<ChildT, Log2Dim>::isValueOff(const Coord& xyz) const
+{
+    const Index n = this->coordToOffset(xyz);
+    if (this->isChildMaskOff(n)) return this->isValueMaskOn(n);
```
Comment (kmuseth) [inline openvdb/openvdb/tree/InternalNode.h:1639]: why not shorten these two lines to:

`return this->isChildMaskOff(n) ? this->isValueMaskOn(n) : mNodes[n].getChild()->isValueOff(xyz);`

File: openvdb/openvdb/tree/InternalNode.h [hunk 3]
```diff
@@ -2272,6 +2358,161 @@ InternalNode<ChildT, Log2Dim>::getLastValue() const
...
+    auto* child = mNodes[n].getChild();
+    mChildMask.setOff(n);
+    mValueMask.set(n, active);
+    mNodes[n].setValue(value);
+    return child;
+}
+
+template<typename ChildT, Index Log2Dim>
+inline void
+InternalNode<ChildT, Log2Dim>::deleteChildUnsafe(Index n, const ValueType& value, bool active)
```
Comment (kmuseth) [inline openvdb/openvdb/tree/InternalNode.h:2501]: in the interest of minimizing code duplication and making it more intuitive this method should be implemented in terms of the previous on:

```
template<typename ChildT, Index Log2Dim>
inline void
InternalNode<ChildT, Log2Dim>::deleteChildUnsafe(Index n, const ValueType& value, bool active)
{
   delete this->stealChildUnsafe(n, value, active);
}
```

File: openvdb/openvdb/tree/LeafNode.h [hunk 4]
```diff
@@ -475,7 +475,11 @@ class LeafNode
     /// Return @c true if the voxel at the given coordinates is active.
     bool isValueOn(const Coord& xyz) const {return this->isValueOn(LeafNode::coordToOffset(xyz));}
     /// Return @c true if the voxel at the given offset is active.
-    bool isValueOn(Index offset) const { return mValueMask.isOn(offset); }
+    bool isValueOn(Index offset) const { OPENVDB_ASSERT(offset < SIZE); return mValueMask.isOn(offset); }
+    /// Return @c true if the voxel at the given coordinates is inactive.
+    bool isValueOff(const Coord& xyz) const {return this->isValueOff(LeafNode::coordToOffset(xyz));}
+    /// Return @c true if the voxel at the given offset is inactive.
+    bool isValueOff(Index offset) const { return mValueMask.isOff(offset); }
```
Comment (kmuseth) [inline openvdb/openvdb/tree/LeafNode.h:482]: missing:

`OPENVDB_ASSERT(offset < SIZE);`

File: openvdb/openvdb/tree/RootNode.h [hunk 5]
```diff
@@ -755,6 +755,35 @@ class RootNode
...
+    //
+    // WARNING: For improved performance, these unsafe methods assume that the tile
+    // or child exists. If used incorrectly, this can cause the application to crash.
+    // Always use the safer alternative method(s) unless you really know what you're doing.
+    // Enabling OpenVDB asserts will catch where assumptions are incorrectly invalidated.
+
+    /// @brief Return the tile value at the given coordinate.
+    /// @note Use cbeginValueAll() for a safer alternative.
+    /// @warning This method should only be used by experts seeking low-level optimizations.
+    const ValueType& getValueUnsafe(const Coord& xyz) const;
```
Comment (kmuseth) [inline openvdb/openvdb/tree/RootNode.h:769]: why is this not call `getTileValueUnsafe` ?

File: openvdb/openvdb/tree/RootNode.h [hunk 6]
```diff
@@ -755,6 +755,35 @@ class RootNode
...
+    /// @brief Return the tile value and active state at the given coordinate.
+    /// @note Use cbeginValueAll() for a safer alternative.
+    /// @warning This method should only be used by experts seeking low-level optimizations.
+    bool getValueUnsafe(const Coord& xyz, ValueType& value) const;
```
Comment (kmuseth) [inline openvdb/openvdb/tree/RootNode.h:773]: why is this not called `getTileValueUnsafe` ?

**Review by kmuseth — CHANGES_REQUESTED** [review]
please see requests for changes

**Comment by danrbailey** [comment]
Addressed all feedback, thanks!

**Review by kmuseth — APPROVED** [review]

**Pull request merged**